Add AWS IAM authorization support

viktor-zinchenko opened the pull request.

Commit:
Add AWS IAM authorization support

mghoneimy commented on composer.json:
```diff
@@ -33,7 +33,8 @@
         "psr/http-client": "^1.0",
-        "guzzlehttp/guzzle": "^6.3|^7.0.1"
+        "guzzlehttp/guzzle": "^6.3|^7.0.1",
+        "aws/aws-sdk-php": "^3.185"
```
I do not want aws-sdk-php to become part of the package dependencies, as it's not required for the functionality of the package, but is rather an extension. Maybe we should it to the `suggest` section instead.

viktor-zinchenko replied:
@mghoneimy Thank you for a feedback. I agree it's not a good idea to add this package to require section as it could be not needed. Then another question: I can't cover the the aws auth with test if I don't have the package installed. So, are we good to add it to require-dev?

mghoneimy commented on src/Client.php:
```diff
@@ -138,4 +163,13 @@ public function runRawQuery(string $queryString, $resultsAsArray = false, array
     }
+
+    /**
+     * @param Request $request
+     * @return Request
+     */
+    public function handleAuth(Request $request): Request
+    {
+        return $this->auth->run($request, $this->httpHeaders);
+    }
```
Why do we need this method?

mghoneimy commented on src/Auth/AwsIamAuth.php:
```diff
@@ -0,0 +1,32 @@
+     */
+    public function run(Request $request, array $headers = []): Request
+    {
+        $region = getenv('AWS_REGION');
```
The package should not assume that the AWS region is declared in the environment variables, this would be bad design. I would suggest passing the AWS region to the constructor of this class as an alternative.

mghoneimy commented on src/Auth/AwsIamAuth.php:
```diff
@@ -0,0 +1,32 @@
+    public function run(Request $request, array $headers = []): Request
+    {
+        $region = getenv('AWS_REGION');
+        $signature = new \Aws\Signature\SignatureV4('appsync', $region);
```
If you composer "recommend" instead of "require" this package, I would check if this class exists before attempting to to instantiate and throw an exception notifying the user to install the package if they want to utilize this feature.

mghoneimy commented on src/Client.php:
```diff
@@ -47,14 +68,20 @@ class Client
         array $httpOptions = [],
         ClientInterface $httpClient = null,
-        string $requestMethod = 'POST'
+        string $requestMethod = 'POST',
+        string $authType = self::AUTH_TYPE_HEADER
     ) {
+        if (!in_array($authType, self::ALLOWED_AUTH_TYPES, true)) {
+            throw new AuthTypeNotSupportedException($requestMethod);
+        }
+        $this->auth = new (self::AUTH_MAP[$authType]);
```
Instead of having the user pass a string to the constructor and then we instantiate a class for it, I would recommend adding a nullable `AuthInterface` argument to the constructor, then the user would pass their own authenticator if they need it.

mghoneimy reviewed (changes requested):
I have left a few comments specific to design and implementation. A more general comment is that none of the tests are updated. We need to update the test classes to verify the modified functionalities.

viktor-zinchenko commented:
@mghoneimy thank you for a review. Fixed the comments, but some things are to be discussed, looking forward.

mghoneimy commented on src/Client.php:
```diff
@@ -32,35 +34,46 @@ class Client
-        string $requestMethod = 'POST'
+        string $requestMethod = 'POST',
+        AuthInterface $auth = null,
+        array $options = []
     ) {
         $headers = array_merge(
             $authorizationHeaders,
             $httpOptions['headers'] ?? [],
             ['Content-Type' => 'application/json']
         );
-
+        $this->auth = $auth ?? new HeaderAuth();
+        $this->options = $options;
+        $this->options['headers'] = $headers;
```
I don't like that we have a new argument for the constructor `options` and that we renamed `httpHeader` class variable to `options` this makes things very complicated and confusing.
Why don't we just have the user pass an extra value to the `httpOptions` array they already use, and make the array key more specific to be `aws_region" instead of just "region"?

viktor-zinchenko replied:
I've changed code in the way you described.
One thing - `aws_region` parameter has nothing to do with $httpOptions, I would say it's just an $option. So, maybe we can rename $httpOptions -> $options to be more consistend? (I would be completely fine if we leave it as it is)

mghoneimy replied:
I don't disagree with you on this. The reason why I want to leave this as it is is because I already have a v2.0 under dev that will overhaul how this works and I want to include your feature in this release. So if we don't modify the current setup a lot, it makes rebasing my work easier. In other words, I'd rather not introduce too many changes unless they're critical. Hope that makes sense.

viktor-zinchenko replied:
I understand, it makes sense.

mghoneimy reviewed (changes requested):
We are very close here. The comments I added are mostly small. I'm trying to figure out why travis builds stopped on build requests too. Once both these things are done, we should be good to go.

viktor-zinchenko commented on composer.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "gmostafa/php-graphql-client",
+    "name": "viktor-zinchenko/php-graphql-client",
```
sorry, that was done for testing in another branch

viktor-zinchenko commented:
Tests are fixed now

mghoneimy commented on src/Client.php:
```diff
@@ -111,16 +126,14 @@ public function runRawQuery(string $queryString, $resultsAsArray = false, array
     {
         $request = new Request($this->requestMethod, $this->endpointUrl);
 
-        foreach($this->httpHeaders as $header => $value) {
-            $request = $request->withHeader($header, $value);
-        }
-
```
I just realized a problem while running some scenarios in my head. So currently, headers are used for multiple purposes other than just authorization, such as passing content type for instance. What is a user sets some headers for their requests in the constructor but they also want to use AWS authorization? With the current implementation, AWS authenticator will takeover and ignore all the headers the user passed for non-authorization purposes, causing regression.

In other words, because headers are used for authorization and non-authorization purposes, when we bypass setting the headers because of AWS authorization, we lose the headers passed by the user for non-authorization purposes.

To solve this problem, I propose one of two solutions:
1. Have the AwsIamAuth extend HeaderAuth, and in the `run()` implementation, execute the `parent::run()` at the beginning to set headers, then pass the returned request for the current logic.
2. Revert this change and have the for loop in all cases, remove the `HeaderAuth` class, and only run the authenticator `run()` if it's not null.

I'm open to any other options you have.

viktor-zinchenko replied:
I understand. So made changes as described in the second option.

mghoneimy reviewed (changes requested):
Thank you for addressing all the previous comments and fixing the tests!

I think that one comment is was not completely addressed and I think that I found a problem worth looking into before we merge this in.

viktor-zinchenko commented on src/Client.php:
```diff
@@ -111,16 +126,14 @@ public function runRawQuery(string $queryString, $resultsAsArray = false, array
         $bodyArray = ['query' => (string) $queryString, 'variables' => $variables];
         $request = $request->withBody(Utils::streamFor(json_encode($bodyArray)));
 
+        $request = $this->auth->run($request, $this->httpHeaders, $this->options);
```
We don't have the `$this->httpOptions` attribute in this class, so I've added `$this->options` to keep parameters.
Removed `$this->httpHeaders` as they aren't needed anymore.

mghoneimy replied:
I apologize, I misspoke. I meant using `this->httpHeaders` instead of `this->options`. With the current implementation `this->options` is just a subset of `this->httpHeaders` so I don't see why we can't just pass the existing class variable instead of having to create a new one and pass instead.
Please correct me if I'm wrong here.
Also, I don't understand what you mean by "Removed $this->httpHeaders as they aren't needed anymore." I can still see the variable being initalized and used.

viktor-zinchenko replied:
I mean the attribute `this->httpHeaders` has **only headers** info, but not all options:
```
        $headers = array_merge(
            $authorizationHeaders,
            $httpOptions['headers'] ?? [],
            ['Content-Type' => 'application/json']
        );
        ...
        $this->httpHeaders = $headers;
```
So, we can't use this->httpHeaders attribute. That's why a new one was added.

---
"Also, I don't understand what you mean by "Removed $this->httpHeaders as they aren't needed anymore." I can still see the variable being initalized and used."
Never mind here, I mean I removed second argument from here:
`$this->auth->run($request, $this->httpHeaders, $this->options);`

mghoneimy replied:
You are right, thanks for clarifying this. I'm good with the changes now.

viktor-zinchenko commented on src/Client.php:
```diff
@@ -34,33 +36,45 @@ class Client
      */
     protected $httpHeaders;
 
+    /**
+     * @var array
+     */
+    protected $options;
+
```
replied above

mghoneimy commented:
Can you rebase your branch on the latest updates on `master` please, I had to replace travis-ci with Github actions to have the tests continue to run.

mghoneimy commented:
Thanks for addressing all the comments. I accepted the changes, but again, I still need need you to rebase your changes on master before we can merge the PR.

viktor-zinchenko commented:
> Thanks for addressing all the comments. I accepted the changes, but again, I still need need you to rebase your changes on master before we can merge the PR.

Great! Thank you.
Yeah, no problem. I've rebased the branch.
Checks were not started, please, run them to ensure everything is fine

longmtran commented:
Hi, does anyone have an example of how the code should look like with this? As in, when making a query to a GraphQL endpoint with IAM auth.
